build-tools/packages/build-cli/src/codeCoverageAnalysisTool/getCoverageMetrics.ts

```diff
@@ -0,0 +1,148 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+/* eslint-disable @typescript-eslint/no-unsafe-assignment */
+/* eslint-disable no-await-in-loop */
+/* eslint-disable @typescript-eslint/dot-notation */
+/* eslint-disable @typescript-eslint/no-unsafe-argument */
+/* eslint-disable @typescript-eslint/no-unsafe-member-access */
+
+import * as fs from "node:fs/promises";
+import * as path from "node:path";
+import globby from "globby";
+import JSZip from "jszip";
+import { Parser } from "xml2js";
+
+/**
+ * The type for the coverage report, containing the name of the package, line coverage and branch coverage
+ */
+export interface CoverageReport {
+	packageName: string;
+	lineCoverage: number;
+	branchCoverage: number;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any -- missing type for XML output
+const extractCoverageMetrics = (result: any): CoverageReport[] => {
+	const report: CoverageReport[] = [];
+	// eslint-disable-next-line @typescript-eslint/no-explicit-any
+	const coverageForPackagesResult = result?.coverage?.packages?.[0]?.package as any[];
+
+	for (const coverageForPackage of coverageForPackagesResult) {
+		const packageName = coverageForPackage["$"].name as string;
+		const lineCoverage = Number.parseFloat(coverageForPackage["$"]["line-rate"]);
+		const branchCoverage = Number.parseFloat(coverageForPackage["$"]["branch-rate"]);
+		report.push({
+			packageName,
+			lineCoverage,
+			branchCoverage,
+		});
+	}
+
+	return report;
+};
+
+/**
+ * Method that returns the coverage report for the baseline build
+ * @param baselineZip - zipped coverage files for the baseline build
+ * @returns an array of coverage metrics for baseline containing packageName, lineCoverage and branchCoverage
+ */
+export const getCoverageMetricsForBaseline = async (
+	baselineZip: JSZip,
+): Promise<CoverageReport[]> => {
+	const coverageReportsFiles: string[] = [];
+	baselineZip.forEach((filePath) => {
+		if (filePath.endsWith("cobertura-coverage-patched.xml"))
+			coverageReportsFiles.push(filePath);
+	});
+
+	const coverageMetricsForBaseline: CoverageReport[] = [];
+	const xmlParser = new Parser();
+
+	try {
+		console.log(`${coverageReportsFiles.length} coverage files found in baseline`);
```

Comment:
In general we don't cosole-log in build-tools. Where needed, take a Logger object as a parameter and call its methods,
or refactor the code to throw exceptions instead.

The general principle is the writing to the console should be done by a CLI command, not arbitrary code running. Passing
a Logger around is a the best approach I've found to keeping the CLI command in control of the logging but not
completely refactoring everything to not require in-function logging.

And when logging with the Logger, prefer verbose unless you KNOW it's useful for every run of the tool.

Reply:
Replaced with the logger.